File: NanoRpc.Net/src/RpcController.cs
namespace NanoRpc
{
    using System;

    /// <summary>
    /// </summary>
    /// TODO: Split this class to RpcClientController and RpcServerController.
    public abstract class RpcController
    {
        /// <summary>
        /// </summary>
        public RpcChannel Channel { get; set; }

        /// <summary>
        /// Gets or sets message recipient.
        /// </summary>
        public IRpcMessageRecipient Recipient { get; set; }

        /// <summary>
        /// </summary>
        /// <param name="rpcMessage">
        /// </param>
        public void Send(RpcMessage rpcMessage)
        {
            CodeContracts.NotNull(Channel, "Channel", new InvalidOperationException("Channel property is null."));

            Channel.Send(rpcMessage);
        }

        /// <summary>
        /// </summary>
        /// <param name="rpcMessage">
        /// </param>
        public abstract void Receive(RpcMessage rpcMessage);
    }
}

File: NanoRpc/src/rpc_controller.cpp

#include "rpc_controller.hpp"
#include "rpc_channel.hpp"
#include "rpc_server.hpp"
#include "rpc_client.hpp"

namespace NanoRpc {

void RpcController::Send(const RpcMessage &message) { channel_->Send(message); }

void RpcController::Receive(const RpcMessage &message) {
  // TODO: Implement client handling
  if (message.has_result() && message.result().status() != RpcSucceeded) {
    if (server_ != NULL)
      server_->Receive(message);
  } else {
    if (message.has_call()) {
      if (server_ != NULL)
        server_->Receive(message);
    }
  }
}

} // namespace
